sdks/python/apache_beam/runners/common.py

```diff
@@ -177,66 +246,37 @@ def old_dofn_process(self, element):
+    self._process_outputs(element, self.dofn.process(element.value))
+
   def new_dofn_process(self, element):
     self.context.set_element(element)
-    arguments, _, _, defaults = self.dofn.get_function_arguments('process')
-    defaults = defaults if defaults else []
-
-    self_in_args = int(self.dofn.is_process_bounded())
 
     # Call for the process function for each window if has windowed side inputs
     # or if the process accesses the window parameter. We can just call it once
     # otherwise as none of the arguments are changing
-    if self.has_windowed_side_inputs or core.NewDoFn.WindowParam in defaults:
+    if self.has_windowed_side_inputs:
       windows = element.windows
     else:
       windows = [window.GlobalWindow()]
 
     for w in windows:
-      args, kwargs = util.insert_values_in_args(
-          self.args, self.kwargs,
-          [s[w] for s in self.side_inputs])
-
-      # If there are more arguments than the default then the first argument
-      # should be the element and the rest should be picked from the side
-      # inputs as window and timestamp should always be tagged
-      if len(arguments) > len(defaults) + self_in_args:
-        if core.NewDoFn.ElementParam not in defaults:
-          args_to_pick = len(arguments) - len(defaults) - 1 - self_in_args
-          final_args = [element.value] + args[:args_to_pick]
-        else:
-          args_to_pick = len(arguments) - len(defaults) - self_in_args
-          final_args = args[:args_to_pick]
+      if self.has_windowed_side_inputs:
+        args, kwargs = util.insert_values_in_args(
+            self.args, self.kwargs, [si[w] for si in self.side_inputs])
       else:
-        args_to_pick = 0
-        final_args = []
-      args = iter(args[args_to_pick:])
-
-      for a, d in zip(arguments[-len(defaults):], defaults):
-        if d == core.NewDoFn.ElementParam:
-          final_args.append(element.value)
-        elif d == core.NewDoFn.ContextParam:
-          final_args.append(self.context)
-        elif d == core.NewDoFn.WindowParam:
-          final_args.append(w)
-        elif d == core.NewDoFn.TimestampParam:
-          final_args.append(element.timestamp)
-        elif d == core.NewDoFn.SideInputParam:
-          # If no more args are present then the value must be passed via kwarg
-          try:
-            final_args.append(args.next())
-          except StopIteration:
-            if a not in kwargs:
-              raise
-        else:
-          # If no more args are present then the value must be passed via kwarg
-          try:
-            final_args.append(args.next())
-          except StopIteration:
-            if a not in kwargs:
-              kwargs[a] = d
-      final_args.extend(list(args))
-      self._process_outputs(element, self.dofn.process(*final_args, **kwargs))
+        args, kwargs = self.args, self.kwargs
+
+      for i, p in self.placeholders:
+        if p == core.NewDoFn.ElementParam:
```

Comment:
That's strange. Maybe it's due to the pickle round trip... Add a TODO to investigate.